concurrency/smallTest/SmallAtomViorOrOverSync.java

```diff
@@ -0,0 +1,190 @@
+package concurrency.smallTest;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * This is for testing messaging, because we have several scenarios that keep interacting one with another
+ * */
+public class SmallAtomViorOrOverSync {
+    private ConcurrentHashMap<String, String> myMap = new ConcurrentHashMap<>();
+
+    String key = "lala" + 33;
+    String val = "lala" + 44;
+    
+    public String give() {
+        myMap.remove("key"); // this remove ensures the noRemove() policy does not kick in
+        return myMap.get("lala");
+    }
+    
+    // ========= check and put =========
+    public void atomViol01() {
+        String myVal = myMap.get(key);
+        if (myVal == null) {
+            myMap.put(key, val);
+        }
+    }
+    
+    public void atomViol02() {
+        if (myMap.get(key) == null) {
+            myMap.put(key, val);
+        }
+    }
+    
+    public void atomViol03() {
+        if ( ! myMap.containsKey(key)) {
+            myMap.put(key, val);
+        }
+    }
+    
+    // =========== check and get ===========
+    
+    public void atomViol04() {
+        String myVal = myMap.get(key);
+        if (myVal != null) {
+            String myVal = myMap.get(key);
+            System.out.println(myVal.length()); // ensure accesses null
+        }
+    }
+    
+    public void atomViol05() {
+        if (myMap.get(key) != null) {
+            String myVal = myMap.get(key);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `get()` may not be thread-safe at lines: **50 and 51**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The removes that can remove the key are at lines: **15**, **67**, and **150**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).